src/components/appointments/DateSelect.js

```diff
@@ -0,0 +1,70 @@
+import {eachDayOfInterval, formatISO, parseISO} from 'date-fns';
+import {useFormikContext} from 'formik';
+import React, {useCallback, useContext} from 'react';
+import {FormattedMessage} from 'react-intl';
+import {useAsync} from 'react-use';
+
+import {ConfigContext} from 'Context';
+import {get} from 'api';
+import {DateField} from 'components/forms';
+
+const getDates = async (baseUrl, productIds, locationId) => {
+  if (!productIds.length || !locationId) return [];
+  const multiParams = productIds.map(id => ({product_id: id}));
+  const datesList = await get(
+    `${baseUrl}appointments/dates`,
+    {location_id: locationId},
+    multiParams
+  );
+  const results = datesList.map(item => item.date);
+  // Array.prototype.toSorted is too new, jest tests can't handle it yet
+  return results.sort();
+};
+
+const DateSelect = () => {
+  const {baseUrl} = useContext(ConfigContext);
+  const {values} = useFormikContext();
+
+  const productIds = (values.products || []).map(prod => prod.product);
+  const locationId = values.location;
+
+  // get the available dates from the API
+  const {
+    loading,
+    value: availableDates = [],
+    error,
+  } = useAsync(
+    async () => await getDates(baseUrl, productIds, locationId),
+    // about JSON.stringify: https://github.com/facebook/react/issues/14476#issuecomment-471199055
+    [baseUrl, JSON.stringify(productIds), locationId]
+  );
+
+  if (error) {
+    throw error;
+  }
+
+  const today = new Date();
+  const minDate = availableDates.length ? parseISO(availableDates[0]) : today;
+  const maxDate = availableDates.length ? parseISO(availableDates.at(-1)) : today;
```

Comment:
The name availableDates suggest a set of dates to me, not a 2-tuple (start, end). I think the interval `[start, end>` is sanest, but we depend on 3rd parties; I know…

If no dates are available, today isn't either, right? Shouldn't that render some other suggestion? Raise an error maybe?

I love https://pypi.org/project/portion/ for these sort of problems.

Reply:
availableDates is a `str[]` with dates in ISO 8601 format, it has arbitrary length. It's not a two-tuple. min and max are set to the first and last possible date, respectively. This makes up the interval of possible available dates, anything before or after is marked as disabled.

The 'gaps' in that array are then explicitly turned into disabled events for the calendar. If no dates are available, min and max are set to the same value and this results in all dates being disabled, which can indeed be improved by just rendering a message "sorry, no available dates, try again later" or something like instead.

I think there's also a bug when only one available date is returned, I'll add a story for that too.

Reply:
> availableDates is a `str[]` with dates in ISO 8601 format, it has arbitrary length. It's not a two-tuple. min and max are set to the first and last possible date, respectively. This makes up the interval of possible available dates, anything before or after is marked as disabled.

Oh right. I remember the api for the date picker.